libqtile/notify.py

```diff
@@ -23,109 +23,128 @@
+
+    class Notification:
+        def __init__(self, summary, body='', timeout=-1, hints=None, app_name='',
+                     replaces_id=None, app_icon=None, actions=None):
+            self.summary = summary
+            self.body = body
+            self.timeout = timeout
+            self.hints = hints or {}
+            self.app_name = app_name
+            self.replaces_id = replaces_id
+            self.app_icon = app_icon
+            self.actions = actions
+
+    class NotificationManager:
+        def __init__(self):
+            self.notifications = []
+            self.callbacks = []
+            self._service = None
+
+        async def service(self):
+            if self._service is None:
+                try:
+                    bus = await MessageBus().connect()
+                    self._service = NotificationService(self)
+                    bus.export(SERVICE_PATH, self._service)
+                    await bus.request_name(BUS_NAME)
+                except Exception:
+                    logger.exception('Dbus connection failed')
+                    self._service = None
+            return self._service
+
+        async def register(self, callback, capabilities=None):
+            service = await self.service()
+            if not service:
+                logger.warning(
+                    'Registering %s without any dbus connection existing',
+                    callback.__name__,
+                )
+            self.callbacks.append(callback)
+            if capabilities:
+                self._service.register_capabilities(capabilities)
+
+        def add(self, notif):
+            self.notifications.append(notif)
+            notif.id = len(self.notifications)
+            for callback in self.callbacks:
+                callback(notif)
+            return len(self.notifications)
+
+        def show(self, *args, **kwargs):
+            notif = Notification(*args, **kwargs)
+            return (notif, self.add(notif))
+
+    notifier = NotificationManager()
+
+else:
+
+    class FakeManager:
+        def __init__(self):
             logger.warning(
-                'Registering %s without any dbus connection existing',
-                callback.__name__,
+                "dbus-next is not installed. "
+                "Notification service and widget are unavailable."
             )
-        self.callbacks.append(callback)
-        if capabilities:
-            self._service.register_capabilities(capabilities)
-
-    def add(self, notif):
-        self.notifications.append(notif)
-        notif.id = len(self.notifications)
-        for callback in self.callbacks:
-            callback(notif)
-        return len(self.notifications)
-
-    def show(self, *args, **kwargs):
-        notif = Notification(*args, **kwargs)
-        return (notif, self.add(notif))
 
+        async def register(self, *args, **kwargs):
+            pass
 
-notifier = NotificationManager()
+    notifier = FakeManager()
```

Comment:
Just to explain this: The `Notify` widget imports `notifier` from this module. In addition, the `Notify` widget does not use `dbus-next`. So, instead of testing whether `dbus-next` is installed in the widget, we import a dummy object that allows the widget to be initialised and configured. A warning message is sent to the log file as well.

Reply:
Might be nice to see if there's another way we can do this (or at least add a comment somewhere with this text in it). However, I don't want to hold this PR up any longer, so we can do it separately.

Reply:
Sure - I'll take a look.

And thanks for merging - looks like you're on a mission today!